Per entity/collider debug rendering

barsoosayque opened the pull request:
I needed a way to toggle debug rendering of colliders for group of entities in my project, so I thought of implementing it kind of like bevy `Wireframe` works, with global flag in resource, and if it's off, then per entity with marker component. I also considered just using `Visibility` for collider entities, but that's seems unintuitive.

~Oh, I also fixed `RapierDebugRenderPlugin::enabled` not being used, I'm not sure if creating context with `enabled: true` was intentional or just a typo.~ (fixed in recent master)

Dessix commented on src/render/mod.rs:
```diff
@@ -131,6 +147,21 @@ impl<'world, 'state, 'a, 'b, 'c> BevyLinesRenderBackend<'world, 'state, 'a, 'b,
+                .get(h)
+                .map(|co| {
+                    let entity = Entity::from_bits(co.user_data as u64);
+                    self.global || self.visible.contains(entity)
```
`self.global` takes precedence above the component, it would seem; Consider early-return in the function with a simple `if self.global { return true; }`, avoiding the rest of this lookup behaviour.

Dessix reviewed (changes requested):
Looks like it would work, but it could use a little bit of cleanup.

Is there any way to elevate "draw only entities with `ColliderDebug` components" to a system, and have a global system and a selective one, and bail out at the start of either, instead of doing this per-line?

barsoosayque commented on src/render/mod.rs:
```diff
@@ -131,6 +147,21 @@ impl<'world, 'state, 'a, 'b, 'c> BevyLinesRenderBackend<'world, 'state, 'a, 'b,
+                    self.global || self.visible.contains(entity)
+                })
+                .unwrap_or(false),
+            _ => true,
```
Well that's what I'm using at the moment, colliders only, so I wasn't even sure if joints are rendered ? And for the bodies, I guess it's only the axes thing that is rendered. But what is DebugRenderObject::Other ?.. And how would I render joints, only if two bodies are marked as debug-visible ? Or any of two bodies is enough to be rendered

barsoosayque commented:
> Is there any way to elevate "draw only entities with ColliderDebug components" to a system, and have a global system and a selective one, and bail out at the start of either, instead of doing this per-line?

I think no, there is no way. Well I don't see any, since rendering is done by `DebugRenderPipeline` which is using whole `RigidBodySet`, `ColliderSet`, etc. So it seems I can't use Queries to filter what to render before calling a `.render` fn, only in the process.

Aceeri commented:
Overall a good idea/impl imo :)

barsoosayque commented:
I have updated this PR once again, I wonder if anything is blocking this PR from being merged ? Would really like to get this feature into the library !

Vrixyz commented:
> I have updated this PR once again, I wonder if anything is blocking this PR from being merged ? Would really like to get this feature into the library !

Thanks for the pull request!

I like it overall, I, too, would have appreciated a more customizable approach, to more easily switching from different modes/selection, maybe https://github.com/bevyengine/bevy/pull/7328 is interesting to take inspiration from it.

In the meanwhile, this is an acceptable implementation for me.

An example and changelog would be great !

barsoosayque commented:
I initially based this PR on how `Wireframe` worked, but that was a long time ago. I'm happy to update it so it would have `AlwaysRender`/`NeverRender` to mimic the modern `Wireframe`, looks straightforward enough. And since I have weekends now, I hope to be done shortly !

barsoosayque commented:
@Vrixyz I've rewrote this PR to mirror bevy's `Wireframe` (hopefully) and added a changelog, a while back actually, just updated the branch. Couldn't add an example, because I'm a little bit burned out from this PR (I just pray that it will be upstreamed this time :pray: )

Vrixyz commented:
Thanks!
